src/mito2/src/compaction/compactor.rs

```diff
@@ -0,0 +1,422 @@
+            compress_type: manifest_compress_type(mito_config.compress_manifest),
+            checkpoint_distance: mito_config.manifest_checkpoint_distance,
+        };
+
+        RegionManifestManager::open(region_manifest_options, Default::default())
+            .await?
+            .context(EmptyRegionDirSnafu {
+                region_id: req.region_id,
+                region_dir,
+            })?
+    };
+
+    let manifest = manifest_manager.manifest();
+    let region_metadata = manifest.metadata.clone();
+    let manifest_ctx = Arc::new(ManifestContext::new(manifest_manager, Writable));
+
+    let file_purger = {
+        let purge_scheduler = Arc::new(LocalScheduler::new(mito_config.max_background_jobs));
+        Arc::new(LocalFilePurger::new(
+            purge_scheduler.clone(),
+            access_layer.clone(),
+            None,
+        ))
+    };
+
+    let version_control = {
+        let memtable_builder = MemtableBuilderProvider::new(None, Arc::new(mito_config.clone()))
+            .builder_for_options(
+                region_options.memtable.as_ref(),
+                !region_options.append_mode,
+            );
+
+        // Initial memtable id is 0.
+        let mutable = Arc::new(TimePartitions::new(
+            region_metadata.clone(),
+            memtable_builder.clone(),
+            0,
+            region_options.compaction.time_window(),
+        ));
+
+        let version = VersionBuilder::new(region_metadata.clone(), mutable)
+            .add_files(file_purger.clone(), manifest.files.values().cloned())
+            .flushed_entry_id(manifest.flushed_entry_id)
+            .flushed_sequence(manifest.flushed_sequence)
+            .truncated_entry_id(manifest.truncated_entry_id)
+            .compaction_time_window(manifest.compaction_time_window)
+            .options(region_options.clone())
+            .build();
+        Arc::new(VersionControl::new(version))
+    };
+
+    Ok(CompactionRegion {
+        region_options: region_options.clone(),
+        manifest_ctx,
+        access_layer,
+        version_control,
+        region_id: req.region_id,
+        cache_manager: Arc::new(CacheManager::default()),
+        engine_config: Arc::new(mito_config.clone()),
+        region_metadata: region_metadata.clone(),
+        file_purger: file_purger.clone(),
+    })
+}
+
+/// MergeOutput represents the output of merging SST files.
+pub struct MergeOutput {
+    pub files_to_add: Option<Vec<FileMeta>>,
+    pub files_to_remove: Option<Vec<FileMeta>>,
+    pub compaction_time_window: Option<i64>,
+}
+
+impl MergeOutput {
+    pub fn is_empty(&self) -> bool {
+        self.files_to_add.is_none() && self.files_to_remove.is_none()
+    }
+}
+
+/// Compactor is the trait that defines the compaction logic.
+#[async_trait::async_trait]
+pub trait Compactor: Send + Sync + 'static {
+    /// Merge SST files for a region.
+    async fn merge_ssts(
+        &self,
+        compaction_region: CompactionRegion,
```

Comment:
Is it possible to take a reference to get rid of one clone?
```suggestion
        compaction_region: &CompactionRegion,
```